serializers/src/main/java/io/pravega/schemaregistry/serializers/AbstractDeserializer.java

```diff
@@ -0,0 +1,122 @@
+    private final SchemaInfo schemaInfo;
+    private final boolean encodeHeader;
+    private final SerializerConfig.Decoders decoders;
+    private final boolean skipHeaders;
+    private final EncodingCache encodingCache;
+
+    protected AbstractDeserializer(String groupId,
+                                   SchemaRegistryClient client,
+                                   @Nullable Schema<T> schema,
+                                   boolean skipHeaders,
+                                   SerializerConfig.Decoders decoders,
+                                   EncodingCache encodingCache,
+                                   boolean encodeHeader) {
+        Preconditions.checkNotNull(groupId);
+        Preconditions.checkNotNull(client);
+        Preconditions.checkNotNull(encodingCache);
+        this.groupId = groupId;
+        this.client = client;
+        this.encodingCache = encodingCache;
+        this.schemaInfo = schema == null ? null : schema.getSchemaInfo();
+        this.encodeHeader = encodeHeader;
+        this.skipHeaders = skipHeaders;
+        this.decoders = decoders;
+            
+        initialize();
+    }
+
+    private void initialize() {
+        if (schemaInfo != null) {
+            log.info("Validate caller supplied schema.");
+            if (!client.canReadUsing(groupId, schemaInfo)) {
+                throw new IllegalArgumentException("Cannot read using schema" + schemaInfo.getType() + " as it is considered incompatible with current policy.");
+            }
+        } else {
+            if (!this.encodeHeader) {
+                log.warn("No reader schema is supplied and stream does not have encoding headers.");
+            }
+        }
+    }
+    
+    @SneakyThrows(IOException.class)
+    @Override
+    public T deserialize(ByteBuffer data) {
+        int start = data.hasArray() ? data.arrayOffset() + data.position() : data.position();
+        if (this.encodeHeader) {
+            SchemaInfo writerSchema = null;
+            ByteBuffer decoded;
+            if (skipHeaders) {
+                data.position(start + HEADER_SIZE);
+                decoded = data;
+            } else {
+                byte protocol = data.get();
+                EncodingId encodingId = new EncodingId(data.getInt());
+                EncodingInfo encodingInfo = encodingCache.getGroupEncodingInfo(encodingId);
+                writerSchema = encodingInfo.getSchemaInfo();
+                decoded = decoders.decode(encodingInfo.getCodecType(), data);
+            }
+
+            ByteArrayInputStream bais = new ByteArrayInputStream(decoded.array(), 
+                    decoded.arrayOffset() + decoded.position(), decoded.remaining());
+            if (schemaInfo == null) { // deserialize into writer schema
+                // pass writer schema for schema to be read into
+                return deserialize(bais, writerSchema, writerSchema);
```

Comment:
These two branches are identical to each other and to the outer else block modulo the assignment of the parameters. It would be easier to follow if the if / else were to assign these and the deserialize call were outside at the end.